test/fixtures/es-module-loaders/mock-loader.mjs

```diff
@@ -0,0 +1,137 @@
+import { receiveMessageOnPort } from 'node:worker_threads';
+const mockedModuleExports = new Map();
+let currentMockVersion = 0;
+
+/**
+ * FIXME: this is a hack to workaround loaders being
+ * single threaded for now
+ */
+function doDrainPort() {
+  let msg;
+  while (msg = receiveMessageOnPort(preloadPort)) {
+    onPreloadPortMessage(msg.message);
+  }
+}
+function onPreloadPortMessage({
+  mockVersion, resolved, exports
+}) {
+  currentMockVersion = mockVersion;
+  mockedModuleExports.set(resolved, exports);
+}
+let preloadPort;
+export function globalPreload({port}) {
+  preloadPort = port;
+  port.on('message', onPreloadPortMessage);
+  port.unref();
+  const insideAppContext = (getBuiltin, port, setImportMetaCallback) => {
+    let mockedModules = new Map();
+    let mockVersion = 0;
+    const doMock = (resolved, replacementProperties) => {
+      let exports = Object.keys(replacementProperties);
+      let namespace = Object.create(null);
+      let listeners = [];
+      for (const name of exports) {
+        let currentValue = replacementProperties[name];
+        Object.defineProperty(namespace, name, {
+          enumerable: true,
+          get() {
+            return currentValue;
+          },
+          set(v) {
+            currentValue = v;
+            for (let fn of listeners) {
+              try {
+                fn(name);
+              } catch {
+              }
+            }
+          }
+        });
+      }
+      mockedModules.set(resolved, {
+        namespace,
+        listeners
+      });
+      mockVersion++;
+      port.postMessage({ mockVersion, resolved, exports });
+      return namespace;
+    }
+    setImportMetaCallback((meta, context, defaultImportMetaInitializer) => {
+      if (context.url === 'node:mock') {
+        meta.doMock = doMock;
+        return;
+      }
+      if (context.url.startsWith('mock:')) {
+        let [proto, version, encodedTargetURL] = context.url.split(':');
+        let decodedTargetURL = decodeURIComponent(encodedTargetURL);
+        if (mockedModules.has(decodedTargetURL)) {
+          meta.mock = mockedModules.get(decodedTargetURL);
+          return;
+        }
+      }
+      defaultImportMetaInitializer(meta, context);
+    });
+  };
+  return `(${insideAppContext})(getBuiltin, port, setImportMetaCallback)`
```

Comment:
Owwww—nice! Much cleaner than both, plus the function declaration stays normal 🙌